### ISSUE TITLE
TypeError: Can't convert 'bytes' object to str implicitly

### ISSUE THREAD

**Issue by sean-hannon** [tweepy/tweepy#615]
I was following the streaming api tutorial and got this error. Other threads say it has been fixed.
Using master branch code.

![typeerror](https://cloud.githubusercontent.com/assets/4067575/7958978/93030500-09ec-11e5-90e9-44732eab1816.PNG)

**Comment by 7diracseas** [comment]
Looks like this just needs to be initialized as ``bytes()`` instead.

**Comment by donummortis** [comment]
Unfortunately, I still get the same error. 
I replaced the line like you suggested and reinstalled, but that just gave me a different error.

Please help me out here, I'm kinda new to this stuff as well :)

**Comment by donummortis** [comment]
fixed it with 2.7 version of python ...

**Comment by numberoverzero** [comment]
Is this being worked on?  Using 2.7 doesn't seem like the right solution when the package is listed as supporting 3.3 and 3.4

**Comment by cozos** [comment]
I changed line `161` to

```
self._buffer += self._stream.read(read_len).decode('ascii')
```

and line `171` to

```
self._buffer += self._stream.read(self._chunk_size).decode('ascii')
```

and then reinstalled.

**Comment by fredrikmollerstrand** [comment]
@cozos' fix works :+1:

**Comment by FlxVctr** [comment]
Thx @cozos. Works for me too.

**Comment by davidchoo12** [comment]
@cozos thanks! fixed for me as well!

**Comment by craighagerman** [comment]
I just got this error too with Python 3.4. The problem seems to be fixed in fix-556 six months ago(!) but hasn't been merged into master.

**Comment by aflyax** [comment]
Same error. Has been working in the past.

**Comment by NicholasNeuteufel** [comment]
Having this issue on a fresh install.

**Comment by alexlu3211** [comment]
@cozos Hey man, that problem was solved, but a new problem came out at the same time...
Traceback (most recent call last):
  File "D:\helloWorld\tweet_classifier.py", line 28, in <module>
    twitterStream.filter(track=["conference"])
  File "C:\Python34\lib\site-packages\tweepy\streaming.py", line 430, in filter
    self._start(async)
  File "C:\Python34\lib\site-packages\tweepy\streaming.py", line 346, in _start
    self._run()
  File "C:\Python34\lib\site-packages\tweepy\streaming.py", line 286, in _run
    raise exception
  File "C:\Python34\lib\site-packages\tweepy\streaming.py", line 255, in _run
    self._read_loop(resp)
  File "C:\Python34\lib\site-packages\tweepy\streaming.py", line 309, in _read_loop
    self._data(next_status_obj)
  File "C:\Python34\lib\site-packages\tweepy\streaming.py", line 289, in _data
    if self.listener.on_data(data) is False:
  File "D:\helloWorld\tweet_classifier.py", line 18, in on_data
    print(tweet)
  File "C:\Python34\lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
UnicodeEncodeError: 'charmap' codec can't encode characters in position 44-47: character maps to <undefined>

Having this issue on UnicodeEncodeError, how could I solve this one? Sorry I'm new to this...thanks for your help in advance!

**Comment by AnbarasuRam** [comment]
@cozos Thank you. Your solution worked.

**Comment by bartmika** [comment]
@cozos This fixes the problem. Thank you!  :+1:

**Comment by Solaxun** [comment]
Silly question maybe.... but after implementing @cozos fix, why do we need to reinstall? Would that not just reload the same broken version?  I implemented the fix to the api file and did not reinstall, so far so good.

**Comment by amirush2001** [comment]
@cozoz, thank you!
JSON format, which Twitter uses, encoded UTF-8. 
So, it is better to change the lines to:
line 161:
self._buffer += self._stream.read(read_len).decode('UTF-8')
line 171:
self._buffer += self._stream.read(self._chunk_size).decode('UTF-8')

**Comment by computermacgyver** [comment]
@Solaxun 
> Silly question maybe.... but after implementing @cozos fix, why do we need to reinstall?

The thought here is that you downloaded a copy of this repository, made the change in the copy of these files on your computer, and then you would install the changed files by running ``python setup.py install`` (or ``python3 setup.py install`` or whatever other command you have for python). This would copy the modified files into the system directories where python can find them.

As I don't believe tweepy has any compiled code it would be fine to find the files in their installed locations and edit them there, in which case reinstalling would obviously be counter productive. That said, I think modifying copies of the files and running the standard install procedure is probably a safer route to go.

**Comment by danielquinn** [comment]
Is there any particular reason that th @cozos fix can't just be merged into core?

**Comment by computermacgyver** [comment]
This has been fixed and proposed in PR #635 (correctly with UTF-8 encoding and backwards compatibility for python 2.x) . It is up to the maintainer to merge that pull request, which I hope they will do quickly.

**Comment by cozos** [comment]
@amirush2001 

Ah, I wasn't sure if I should use UTF-8 or Ascii but I suppose the answer is always UTF-8 these days. I suppose I'll amend my comment since people are using it, sorry :)

**Comment by craigderington** [comment]
@amirush2001 your fix worked for me...

**Comment by chisaipete** [comment]
Yeah, I'm not thrilled that installing tweepy fresh through pip (3.4.0) in python 3.5 is broken and the readme says it supports 3.5.

**Comment by danielquinn** [comment]
@chisaipete sounds like there's a bug somewhere.  Perhaps rather than passive-aggressively complaining about it, you can submit a bug report or pull request to fix it?  This is a Free software project supported by people who do this on their own time.  Please don't shit on them for contributing.

**Comment by chisaipete** [comment]
I wasn't intending to put anyone down, I was simply commenting that the accepted fix which closed this issue is inadequate and the readme is inaccurate.  I suppose I should have been more explicit and diplomatic, my apologies.

**Comment by wjt** [comment]
@chisaipete pretty sure the fix works – but there hasn't been a new release with it, so you'll need to install from git rather than from pypi.

**Comment by joshthecoder** [comment]
3.5.0 should have the fix

**Comment by Philipduerholt** [comment]
Hello everyone. I tryed to fix the lines on streaming.py and - after that didn't work - installed 3.5.0 from github. I still get the Type error: Can't convert 'bytes' object to str implicitly
I use Python 3.4

**Comment by nitishk879** [comment]
I know it's too late but, I hope this will work for upcoming users as well. I'm using python 3+ and I got similar problem when I was using following function
        def __str__(self):
        data = ("Name:" + self.name.encode('UTF-8') + "\r\n Developer: " + self.developer.encode('UTF-8') +
                "\r\n Price: " + self.price.encode('UTF-8') + "\r\n")
        return data
I made a few changes and it work perfectly for me.
    def __str__(self):
        data = ("Name:" + self.name.strip() + "\r\n Developer: " + self.developer.strip() +
                "\r\n Price: " + self.price.strip() + "\r\n")
        return data

**Comment by Harmon758** [comment]
This should be fixed with https://github.com/tweepy/tweepy/commit/22c0bc029546478da50c46100ad3804863138158 (#635), released with Tweepy v3.5.0.
Feel free to open a new issue if you can reproduce this with the latest development version.

@sean-hannon I'm surprised that no one has told you after all this time, but you're going to want to revoke and regenerate your keys and tokens if you haven't already.

@alexlu3211 This is an issue with your console's code page/encoding rather than an issue with Tweepy.
You're attempting to print a character outside the range that [CP-1252](https://en.wikipedia.org/wiki/Windows-1252) supports.
This is easily resolved by changing the active code page to UTF-8 using [`chcp`](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/chcp) and specifying [`65001`](https://docs.microsoft.com/en-us/windows/desktop/Intl/code-page-identifiers).

**Comment by sean-hannon** [comment]
@Harmon758 I think I gave up and just went and used Twitter4J. It was for a college project so just wanted something that worked. Cheers for replying though.
